File: App/BNKScopeField/Views/ExporterPanel.swift
import SwiftUI
import BNKKit

/// The exporter's state, and what can be done about it, on the screen that
/// depends on it.
///
/// This used to be a separate Telemetry item in the menu, which put a thing you
/// configure beside seven things you look at. Worse, it split one subject in
/// two: TMM Live listed which pods carried the exporter and could do nothing,
/// while another screen could act but showed no graphs. The state and the
/// actions belong together, on the screen that is empty without them.
struct ExporterPanel: View {
    enum Style {
        /// The full targets list, shown under a working dashboard.
        case card
        /// The whole screen, when there is nothing to scrape yet.
        case prompt
    }

    let style: Style
    @Environment(ClusterStore.self) private var store
    @Environment(TelemetryEngine.self) private var engine

    @State private var busy = false
    @State private var confirmingRemoval = false
    @State private var typed = ""
    @State private var owners: [String] = []
    @State private var report: Report?

    struct Report: Identifiable {
        let id = UUID()
        let title: String
        let lines: [String]
        let bad: Bool
    }

    private var pods: [K8s.Pod] { store.current?.tmmPods ?? [] }
    private var missing: [K8s.Pod] { pods.filter { Exporter.installation(in: $0) == .absent } }
    private var ephemeral: [K8s.Pod] { pods.filter { Exporter.installation(in: $0) == .ephemeral } }
    private var permanent: [K8s.Pod] {
        pods.filter { if case .permanent = Exporter.installation(in: $0) { return true } else { return false } }
    }

    var body: some View {
        Group {
            switch style {
            case .card:   targetsCard
            case .prompt: installPrompt
            }
        }
        .task(id: "\(store.selected ?? "")#\(store.current?.probeGeneration ?? 0)") { await findOwners() }
        .alert("Recreate the TMM pods?", isPresented: $confirmingRemoval) {
            TextField("cluster name", text: $typed)
                .textInputAutocapitalization(.never).autocorrectionDisabled()
            Button("Cancel", role: .cancel) { typed = "" }
            Button("Recreate pods and remove", role: .destructive) { Task { await remove() } }
                .disabled(typed != store.current?.displayName)
        } message: {
            Text("An ephemeral container cannot be taken out of a running pod. Removing the exporter means deleting \(ephemeral.count) f5-tmm pod\(ephemeral.count == 1 ? "" : "s") so they are rebuilt without it. Dataplane traffic through those pods stops until they are back.\n\nType \(store.current?.displayName ?? "the cluster name") to confirm.")
        }
        .sheet(item: $report) { r in
            VStack(alignment: .leading, spacing: 14) {
                Text(r.title).font(.system(size: 17, weight: .semibold)).foregroundStyle(Theme.fg)
                ForEach(r.lines, id: \.self) { line in
                    Text(line).font(Theme.mono(12.5))
                        .foregroundStyle(r.bad ? Theme.warn : Theme.muted)
                        .fixedSize(horizontal: false, vertical: true)
                }
                Spacer()
                Button("Done") { report = nil }.buttonStyle(.borderedProminent)
            }
            .padding(22).frame(minWidth: 420, minHeight: 220).background(Theme.bg)
        }
    }

    // MARK: - The empty case

    private var installPrompt: some View {
        VStack(spacing: 14) {
            Text("Nothing to scrape here")
                .font(.system(size: 17, weight: .semibold)).foregroundStyle(Theme.fg)
            Text("\(pods.count) f5-tmm pod\(pods.count == 1 ? "" : "s") on this cluster, none carrying the exporter. Adding it attaches an ephemeral container that reads the tmstat segment read-only and serves /metrics. TMM keeps running — nothing restarts.")
                .font(.system(size: 13)).foregroundStyle(Theme.muted)
                .multilineTextAlignment(.center).frame(maxWidth: 470)
            Button(busy ? "Adding…" : "Add the exporter") { Task { await install() } }
                .buttonStyle(.borderedProminent).disabled(busy)
            Text("It is gone again when a pod is recreated, and nothing re-adds it.")
                .font(.system(size: 11.5)).foregroundStyle(Theme.faint)
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .background(Theme.bg)
    }

    // MARK: - The working case

    private var targetsCard: some View {
        VStack(alignment: .leading, spacing: 10) {
            HStack(spacing: 10) {
                Text("Exporter targets").font(.system(size: 13.5, weight: .semibold)).foregroundStyle(Theme.fg)
                Text("tmm-stat-exporter").font(Theme.mono(11)).foregroundStyle(Theme.muted)
                Spacer(minLength: 8)
                actions
            }
            ForEach(pods, id: \.metadata.name) { pod in
                TargetRow(pod: pod, status: engine.podStatus[pod.metadata.name])
            }
            if !permanent.isEmpty {
                Text(owners.isEmpty
                     ? "Defined in the workload's pod template, so this app cannot remove it."
                     : "Defined in \(owners.joined(separator: ", ")) — removing it means editing that, not deleting pods.")
                    .font(.system(size: 11.5)).foregroundStyle(Theme.faint)
                    .fixedSize(horizontal: false, vertical: true)
            }
        }
        .padding(14)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(Theme.card, in: RoundedRectangle(cornerRadius: 12))
        .overlay(RoundedRectangle(cornerRadius: 12).strokeBorder(Theme.border))
    }

    @ViewBuilder
    private var actions: some View {
        if busy {
            ProgressView().controlSize(.small)
        } else {
            if !missing.isEmpty {
                Button("Add to \(missing.count) pod\(missing.count == 1 ? "" : "s")") { Task { await install() } }
                    .buttonStyle(.borderedProminent).controlSize(.small)
            }
            if !ephemeral.isEmpty {
                Button("Remove…") { typed = ""; confirmingRemoval = true }
                    .buttonStyle(.bordered).controlSize(.small).tint(Theme.bad)
            }
        }
    }

    // MARK: - Doing it

    private func install() async {
        guard let cluster = store.current, let client = try? cluster.client() else { return }
        busy = true
        let outcome = await Exporter.install(into: missing, clusterLabel: cluster.displayName, using: client)
        await cluster.probe()
        busy = false
        report = Report(title: outcome.failed.isEmpty ? "Exporter added" : "Exporter partly added",
                        lines: summarise(outcome, verb: "added to"), bad: !outcome.failed.isEmpty)
    }

    private func remove() async {
        guard let cluster = store.current, let client = try? cluster.client() else { return }
        busy = true
        typed = ""
        engine.stop()
        let outcome = await Exporter.remove(from: ephemeral, using: client)
        await cluster.probe()
        busy = false
        report = Report(title: outcome.failed.isEmpty ? "Pods recreated" : "Removal incomplete",
                        lines: summarise(outcome, verb: "recreated"), bad: !outcome.failed.isEmpty)
    }

    private func findOwners() async {
        owners = []
        guard let cluster = store.current, let client = try? cluster.client() else { return }
        var found: Set<String> = []
        for pod in permanent {
            if let owner = await Exporter.owner(of: pod, using: client) { found.insert(owner) }
        }
        owners = found.sorted()
    }

    private func summarise(_ outcome: Exporter.Outcome, verb: String) -> [String] {
        var lines: [String] = []
        if !outcome.changed.isEmpty {
            lines.append("\(verb) \(outcome.changed.count) pod(s):\n  " + outcome.changed.joined(separator: "\n  "))
        }
        if !outcome.skipped.isEmpty { lines.append("skipped \(outcome.skipped.count), already as wanted") }
        for (pod, reason) in outcome.failed { lines.append("\(pod): \(reason)") }
        return lines.isEmpty ? ["Nothing changed."] : lines
    }
}

/// One exporter target: whether it is answering, and how it got there.
private struct TargetRow: View {
    let pod: K8s.Pod
    let status: TelemetryEngine.PodStatus?

    var body: some View {
        HStack(spacing: 12) {
            StatusDot(color: dot, glow: { if case .answering = status { true } else { false } }())
            VStack(alignment: .leading, spacing: 3) {
                Text(pod.metadata.name).font(Theme.mono(12)).foregroundStyle(Theme.fg)
                    .lineLimit(1).truncationMode(.middle)
                Text(detail).font(Theme.mono(10.5))
                    .foregroundStyle({ if case .failing = status { Theme.warn } else { Theme.faint } }())
                    .lineLimit(1).truncationMode(.tail)
            }
            Spacer(minLength: 8)
            Text(pod.node).font(Theme.mono(11.5)).foregroundStyle(Theme.muted)
                .lineLimit(1).truncationMode(.middle)
            if Exporter.installation(in: pod) == .ephemeral {
                Badge(text: "ephemeral", color: Theme.warn)
            }
        }
        .padding(.horizontal, 14).padding(.vertical, 10)
        .background(Theme.bg, in: RoundedRectangle(cornerRadius: 9))
        .overlay(RoundedRectangle(cornerRadius: 9).strokeBorder(Theme.border))
    }

    private var dot: Color {
        switch status {
        case .answering: return Theme.ok
        case .failing:   return Theme.bad
        case nil:        return Exporter.installation(in: pod) == .absent ? Theme.warn : Theme.muted
        }
    }

    private var detail: String {
        switch status {
        case .answering(let n):
            return "\(n) samples · \(Exporter.runningImage(in: pod) ?? "exporter")"
        case .failing(let why):
            return why
        case nil:
            return Exporter.installation(in: pod) == .absent
                ? "no exporter in this pod"
                : (Exporter.runningImage(in: pod) ?? "exporter present, not scraped yet")
        }
    }
}

File: dart/bnkscope_field/lib/screens/exporter_panel.dart
import 'package:bnk_engines/bnk_engines.dart';
import 'package:bnk_kit/bnk_kit.dart';
import 'package:flutter/material.dart';

import '../engines.dart';
import '../observe.dart';
import '../theme.dart';
import '../widgets.dart';

enum ExporterStyle {
  /// The full targets list, shown under a working dashboard.
  card,

  /// The whole screen, when there is nothing to scrape yet.
  prompt,
}

/// The exporter's state, and what can be done about it, on the screen that
/// depends on it.
///
/// This used to be a separate Telemetry item in the menu, which put a thing
/// you configure beside seven things you look at. Worse, it split one
/// subject in two: TMM Live listed which pods carried the exporter and could
/// do nothing, while another screen could act but showed no graphs. The
/// state and the actions belong together, on the screen that is empty
/// without them.
class ExporterPanel extends StatefulWidget {
  final ExporterStyle style;
  const ExporterPanel({super.key, required this.style});

  @override
  State<ExporterPanel> createState() => _ExporterPanelState();
}

class _ExporterPanelState extends State<ExporterPanel> {
  bool _busy = false;
  List<String> _owners = const [];
  String? _ownersFor;

  List<Pod> _pods(ClusterStore store) => store.current?.tmmPods ?? const [];
  List<Pod> _missing(ClusterStore store) =>
      _pods(store).where((p) => Exporter.installation(p) is AbsentInstallation).toList();
  List<Pod> _ephemeral(ClusterStore store) =>
      _pods(store).where((p) => Exporter.installation(p) is EphemeralInstallation).toList();
  List<Pod> _permanent(ClusterStore store) =>
      _pods(store).where((p) => Exporter.installation(p) is PermanentInstallation).toList();

  @override
  Widget build(BuildContext context) {
    final engines = Engines.of(context);
    final store = engines.store;
    final cluster = store.current;
    return Observe([store, engines.telemetry, if (cluster != null) cluster], builder: (context) {
      final key = '${store.selected ?? ''}#${cluster?.probeGeneration ?? 0}';
      if (_ownersFor != key) {
        _ownersFor = key;
        WidgetsBinding.instance.addPostFrameCallback((_) => _findOwners(store));
      }
      return switch (widget.style) {
        ExporterStyle.card => _targetsCard(context, store, engines.telemetry),
        ExporterStyle.prompt => _installPrompt(context, store),
      };
    });
  }

  // The empty case

  Widget _installPrompt(BuildContext context, ClusterStore store) {
    final pods = _pods(store);
    final missing = _missing(store);
    return Center(
      child: ConstrainedBox(
        constraints: const BoxConstraints(maxWidth: 470),
        child: Column(mainAxisSize: MainAxisSize.min, children: [
          Text(missing.isEmpty ? 'Exporter added' : 'Nothing to scrape here', style: Tokens.text(17, weight: FontWeight.w600)),
          const SizedBox(height: 14),
          Text(
            missing.isEmpty
                ? 'The exporter is in. TMM Live is starting up — the first samples arrive within a scrape or two.'
                : '${missing.length} of ${pods.length} f5-tmm pod${pods.length == 1 ? '' : 's'} on this cluster carry no exporter. '
                    'Adding it attaches an ephemeral container that reads the tmstat segment read-only and serves /metrics. '
                    'TMM keeps running — nothing restarts.',
            style: Tokens.text(13, color: Tokens.muted),
            textAlign: TextAlign.center,
          ),
          const SizedBox(height: 14),
          if (missing.isNotEmpty)
            FilledButton(onPressed: _busy ? null : () => _install(store), child: Text(_busy ? 'Adding…' : 'Add the exporter'))
          else
            const SizedBox(width: 18, height: 18, child: CircularProgressIndicator(strokeWidth: 2)),
          const SizedBox(height: 14),
          Text('It is gone again when a pod is recreated, and nothing re-adds it.', style: Tokens.text(11.5, color: Tokens.faint)),
        ]),
      ),
    );
  }

  // The working case

  Widget _targetsCard(BuildContext context, ClusterStore store, TelemetryEngine engine) {
    final missing = _missing(store);
    final ephemeral = _ephemeral(store);
    final permanent = _permanent(store);
    return Panel(
      padding: const EdgeInsets.all(14),
      child: Column(crossAxisAlignment: CrossAxisAlignment.start, children: [
        Row(children: [
          Text('Exporter targets', style: Tokens.text(13.5, weight: FontWeight.w600)),
          const SizedBox(width: 10),
          Text(Exporter.containerName, style: Tokens.mono(11, color: Tokens.muted)),
          const Spacer(),
          if (_busy)
            const SizedBox(width: 14, height: 14, child: CircularProgressIndicator(strokeWidth: 2))
          else ...[
            if (missing.isNotEmpty)
              FilledButton(
                onPressed: () => _install(store),
                child: Text('Add to ${missing.length} pod${missing.length == 1 ? '' : 's'}'),
              ),
            if (ephemeral.isNotEmpty) ...[
              const SizedBox(width: 8),
              OutlinedButton(
                style: OutlinedButton.styleFrom(foregroundColor: Tokens.bad, side: BorderSide(color: Tokens.bad.withValues(alpha: 0.4))),
                onPressed: () => _confirmRemoval(context, store, engine, ephemeral),
                child: const Text('Remove…'),
              ),
            ],
          ],
        ]),
        for (final pod in _pods(store)) ...[
          const SizedBox(height: 10),
          _TargetRow(pod: pod, status: engine.podStatus[pod.metadata.name]),
        ],
        if (permanent.isNotEmpty) ...[
          const SizedBox(height: 10),
          Text(
            _owners.isEmpty
                ? "Defined in the workload's pod template, so this app cannot remove it."
                : 'Defined in ${_owners.join(', ')} — removing it means editing that, not deleting pods.',
            style: Tokens.text(11.5, color: Tokens.faint),
          ),
        ],
      ]),
    );
  }

  // Doing it

  Future<void> _install(ClusterStore store) async {
    final cluster = store.current;
    if (cluster == null) return;
    final KubeClient client;
    try {
      client = cluster.client();
    } catch (e) {
      // Returning quietly here is indistinguishable from a tap that did
      // nothing, which is what a silent failure looks like from the far side
      // of the screen.
      await _report('Could not reach the cluster', [brief(e)], bad: true);
      return;
    }
    setState(() => _busy = true);
    final outcome = await Exporter.install(_missing(store), clusterLabel: cluster.displayName, client: client);
    await cluster.probeReflectingChange();
    if (mounted) setState(() => _busy = false);
    await _report(outcome.failed.isEmpty ? 'Exporter added' : 'Exporter partly added', _summarise(outcome, 'added to'),
        bad: outcome.failed.isNotEmpty);
  }

  Future<void> _confirmRemoval(BuildContext context, ClusterStore store, TelemetryEngine engine, List<Pod> ephemeral) async {
    final name = store.current?.displayName ?? 'the cluster name';
    final typed = TextEditingController();
    final confirmed = await showDialog<bool>(
      context: context,
      builder: (context) => StatefulBuilder(builder: (context, setDialogState) {
        return AlertDialog(
          title: const Text('Recreate the TMM pods?'),
          content: Column(mainAxisSize: MainAxisSize.min, crossAxisAlignment: CrossAxisAlignment.start, children: [
            Text('An ephemeral container cannot be taken out of a running pod. Removing the exporter means deleting '
                '${ephemeral.length} f5-tmm pod${ephemeral.length == 1 ? '' : 's'} so they are rebuilt without it. '
                'Dataplane traffic through those pods stops until they are back.\n\nType $name to confirm.'),
            const SizedBox(height: 12),
            TextField(
              controller: typed,
              autocorrect: false,
              enableSuggestions: false,
              decoration: const InputDecoration(hintText: 'cluster name', isDense: true, border: OutlineInputBorder()),
              style: Tokens.mono(13),
              onChanged: (_) => setDialogState(() {}),
            ),
          ]),
          actions: [
            TextButton(onPressed: () => Navigator.of(context).pop(false), child: const Text('Cancel')),
            TextButton(
              style: TextButton.styleFrom(foregroundColor: Tokens.bad),
              onPressed: typed.text == name ? () => Navigator.of(context).pop(true) : null,
              child: const Text('Recreate pods and remove'),
            ),
          ],
        );
      }),
    );
    typed.dispose();
    if (confirmed == true) await _remove(store, engine, ephemeral);
  }

  Future<void> _remove(ClusterStore store, TelemetryEngine engine, List<Pod> ephemeral) async {
    final cluster = store.current;
    if (cluster == null) return;
    final KubeClient client;
    try {
      client = cluster.client();
    } catch (e) {
      await _report('Could not reach the cluster', [brief(e)], bad: true);
      return;
    }
    setState(() => _busy = true);
    engine.stop();
    final outcome = await Exporter.remove(ephemeral, client: client);
    await cluster.probeReflectingChange();
    if (mounted) setState(() => _busy = false);
    await _report(outcome.failed.isEmpty ? 'Pods recreated' : 'Removal incomplete', _summarise(outcome, 'recreated'),
        bad: outcome.failed.isNotEmpty);
  }

  Future<void> _findOwners(ClusterStore store) async {
    final cluster = store.current;
    final client = cluster?.clientOrNull;
    if (cluster == null || client == null) {
      if (mounted) setState(() => _owners = const []);
      return;
    }
    final found = <String>{};
    for (final pod in _permanent(store)) {
      final owner = await Exporter.owner(pod, client);
      if (owner != null) found.add(owner);
    }
    if (mounted) setState(() => _owners = found.toList()..sort());
  }

  List<String> _summarise(ExporterOutcome outcome, String verb) {
    final lines = <String>[];
    if (outcome.changed.isNotEmpty) {
      lines.add('$verb ${outcome.changed.length} pod(s):\n  ${outcome.changed.join('\n  ')}');
    }
    if (outcome.skipped.isNotEmpty) lines.add('skipped ${outcome.skipped.length}, already as wanted');
    for (final f in outcome.failed) {
      lines.add('${f.pod}: ${f.reason}');
    }
    return lines.isEmpty ? ['Nothing changed.'] : lines;
  }

  Future<void> _report(String title, List<String> lines, {required bool bad}) async {
    if (!mounted) return;
    await showDialog<void>(
      context: context,
      builder: (context) => AlertDialog(
        title: Text(title),
        content: ConstrainedBox(
          constraints: const BoxConstraints(minWidth: 420),
          child: Column(mainAxisSize: MainAxisSize.min, crossAxisAlignment: CrossAxisAlignment.start, children: [
            for (final line in lines)
              Padding(
                padding: const EdgeInsets.only(bottom: 8),
                child: Text(line, style: Tokens.mono(12.5, color: bad ? Tokens.warn : Tokens.muted)),
              ),
          ]),
        ),
        actions: [FilledButton(onPressed: () => Navigator.of(context).pop(), child: const Text('Done'))],
      ),
    );
  }
}

/// One exporter target: whether it is answering, and how it got there.
class _TargetRow extends StatelessWidget {
  final Pod pod;
  final ScrapeStatus? status;
  const _TargetRow({required this.pod, required this.status});

  @override
  Widget build(BuildContext context) {
    final installation = Exporter.installation(pod);
    final absent = installation is AbsentInstallation;
    final dot = switch (status) {
      Answering() => Tokens.ok,
      Failing() => Tokens.bad,
      null => absent ? Tokens.warn : Tokens.muted,
    };
    final detail = switch (status) {
      Answering(:final samples) => '$samples samples · ${Exporter.runningImage(pod) ?? 'exporter'}',
      Failing(:final why) => why,
      null => absent ? 'no exporter in this pod' : (Exporter.runningImage(pod) ?? 'exporter present, not scraped yet'),
    };
    return Inset(
      child: Row(children: [
        StatusDot(color: dot, glow: status is Answering),
        const SizedBox(width: 12),
        Expanded(
          child: Column(crossAxisAlignment: CrossAxisAlignment.start, children: [
            Text(pod.metadata.name, style: Tokens.mono(12), maxLines: 1, overflow: TextOverflow.ellipsis),
            const SizedBox(height: 3),
            Text(detail, style: Tokens.mono(10.5, color: status is Failing ? Tokens.warn : Tokens.faint), maxLines: 1, overflow: TextOverflow.ellipsis),
          ]),
        ),
        const SizedBox(width: 8),
        Text(pod.node, style: Tokens.mono(11.5, color: Tokens.muted), maxLines: 1, overflow: TextOverflow.ellipsis),
        if (installation is EphemeralInstallation) ...[
          const SizedBox(width: 8),
          const Pill('ephemeral', color: Tokens.warn),
        ],
      ]),
    );
  }
}
